aten/src/ATen/native/Sorting.cpp

```diff
@@ -271,8 +267,124 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_impl_cpu(
   return std::forward_as_tuple(values, indices);
 }
 
+std::tuple<Tensor&, Tensor&> median_out_impl(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t dim,
+    bool keepdim,
+    bool ignore_nan) {
+  NoNamesGuard guard;
+
+  dim = at::maybe_wrap_dim(dim, self.dim());
+  TORCH_CHECK(self.numel() > 0, "median() input tensor cannot be empty");
+  checkDeviceType("median", {values, indices}, self.device().type());
+  checkScalarType("median", {indices, "indices", 1}, kLong);
+  checkSameType("median", {values, "values", 0}, {self, "self", 2});
+
+  std::vector<int64_t> out_shape = self.sizes().vec();
+  if (self.dim() > 0) {
+    if (keepdim) {
+      out_shape[dim] = 1;
+    } else {
+      out_shape.erase(out_shape.begin() + dim);
+    }
+  }
+
+  resize_output(values, out_shape);
+  resize_output(indices, out_shape);
+
+  Tensor vals = keepdim ? values : values.unsqueeze(dim);
+  Tensor inds = keepdim ? indices : indices.unsqueeze(dim);
+
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "median_out", [&] {
+    dim_apply({self, vals, inds}, dim, [ignore_nan](int64_t it, TensorList tl) {
+      Tensor in = tl[0].contiguous();
+      scalar_t* ip = in.data_ptr<scalar_t>();
+      int64_t size = in.numel();
+
+      if (!ignore_nan) {
+        scalar_t* nanp = std::find_if(ip, ip + size, _isnan<scalar_t>);
+        if (nanp != ip + size) {
+          *tl[1].data_ptr<scalar_t>() = *nanp;
+          *tl[2].data_ptr<int64_t>() = nanp - ip;
+          return;
+        }
+      }
+
+      std::vector<int64_t> idx(size);
+      auto first = idx.begin();
+      auto last = idx.end();
+      std::iota(first, last, 0);
+
+      auto nth = first;
+      if (!ignore_nan) {
+        nth += (size - 1) / 2;
+        std::nth_element(first, nth, last, [&ip](int64_t i, int64_t j) {
+          return ip[i] < ip[j] || (ip[i] == ip[j] && i < j);
+        });
+      } else {
+        int64_t num_nan = std::count_if(ip, ip + size, _isnan<scalar_t>);
+        nth += (size - num_nan - 1) / 2;
+        std::nth_element(first, nth, last, [&ip](int64_t i, int64_t j) {
+          return ip[i] < ip[j] || (ip[i] == ip[j] && i < j) ||
+              (_isnan(ip[j]) && !_isnan(ip[i]));
+        });
+      }
+
+      *tl[1].data_ptr<scalar_t>() = ip[*nth];
+      *tl[2].data_ptr<int64_t>() = *nth;
+    });
+  });
+
+  guard.reset();
+  namedinference::propagate_names_for_reduction(values, self, dim, keepdim);
+  namedinference::propagate_names_for_reduction(indices, self, dim, keepdim);
+
+  return std::forward_as_tuple(values, indices);
+}
+
+Tensor median_impl(const Tensor& self, bool ignore_nan) {
+  NoNamesGuard guard;
+
+  int64_t size = self.numel();
+  TORCH_CHECK(size > 0, "median() input tensor cannot be empty");
+
+  Tensor in = self.clone();
+  Tensor out = at::empty({}, self.options());
```

Comment:
`at::empty({}, self.options());`

or 

`at::empty({0}, self.options());`

?

Reply:
We want to return a scalar tensor so the first version is right.